// @backstage/frontend-plugin-api/dist/wiring/createFrontendPlugin.esm.js
import { resolveExtensionDefinition } from './resolveExtensionDefinition.esm.js';

function createFrontendPlugin(options) {
  const extensions = new Array();
  const extensionDefinitionsById = /* @__PURE__ */ new Map();
  for (const def of options.extensions ?? []) {
    const ext = resolveExtensionDefinition(def, { namespace: options.id });
    extensions.push(ext);
    extensionDefinitionsById.set(ext.id, { ...def, namespace: options.id });
  }
  if (extensions.length !== extensionDefinitionsById.size) {
    const extensionIds = extensions.map((e) => e.id);
    const duplicates = Array.from(
      new Set(
        extensionIds.filter((id, index) => extensionIds.indexOf(id) !== index)
      )
    );
    throw new Error(
      `Plugin '${options.id}' provided duplicate extensions: ${duplicates.join(
        ", "
      )}`
    );
  }
  return {
    $$type: "@backstage/BackstagePlugin",
    version: "v1",
    id: options.id,
    routes: options.routes ?? {},
    externalRoutes: options.externalRoutes ?? {},
    featureFlags: options.featureFlags ?? [],
    extensions,
    getExtension(id) {
      return extensionDefinitionsById.get(id);
    },
    toString() {
      return `Plugin{id=${options.id}}`;
    },
    withOverrides(overrides) {
      const overriddenExtensionIds = new Set(
        overrides.extensions.map(
          (e) => resolveExtensionDefinition(e, { namespace: options.id }).id
        )
      );
      const nonOverriddenExtensions = (options.extensions ?? []).filter(
        (e) => !overriddenExtensionIds.has(
          resolveExtensionDefinition(e, { namespace: options.id }).id
        )
      );
      return createFrontendPlugin({
        ...options,
        extensions: [...nonOverriddenExtensions, ...overrides.extensions]
      });
    }
  };
}
const createPlugin = createFrontendPlugin;

export { createFrontendPlugin, createPlugin };
