// @backstage/plugin-catalog-react/dist/alpha/blueprints/EntityCardBlueprint.esm.js
import { createExtensionBlueprint, coreExtensionData, ExtensionBoundary } from '@backstage/frontend-plugin-api';
import { catalogExtensionData } from '../extensions.esm.js';

const EntityCardBlueprint = createExtensionBlueprint({
  kind: "entity-card",
  attachTo: { id: "entity-content:catalog/overview", input: "cards" },
  output: [
    coreExtensionData.reactElement,
    catalogExtensionData.entityFilterFunction.optional(),
    catalogExtensionData.entityFilterExpression.optional()
  ],
  dataRefs: {
    filterFunction: catalogExtensionData.entityFilterFunction,
    filterExpression: catalogExtensionData.entityFilterExpression
  },
  config: {
    schema: {
      filter: (z) => z.string().optional()
    }
  },
  *factory({
    loader,
    filter
  }, { node, config }) {
    yield coreExtensionData.reactElement(ExtensionBoundary.lazy(node, loader));
    if (config.filter) {
      yield catalogExtensionData.entityFilterExpression(config.filter);
    } else if (typeof filter === "string") {
      yield catalogExtensionData.entityFilterExpression(filter);
    } else if (typeof filter === "function") {
      yield catalogExtensionData.entityFilterFunction(filter);
    }
  }
});

export { EntityCardBlueprint };
