// @backstage/frontend-plugin-api/dist/extensions/createRouterExtension.esm.js
import React from 'react';
import { createExtension } from '../wiring/createExtension.esm.js';
import { RouterBlueprint } from '../blueprints/RouterBlueprint.esm.js';

function createRouterExtension(options) {
  return createExtension({
    kind: "app-router-component",
    namespace: options.namespace,
    name: options.name,
    attachTo: options.attachTo ?? { id: "app/root", input: "router" },
    configSchema: options.configSchema,
    disabled: options.disabled,
    inputs: options.inputs,
    output: {
      component: RouterBlueprint.dataRefs.component
    },
    factory({ inputs, config }) {
      const Component = (props) => {
        return /* @__PURE__ */ React.createElement(options.Component, { inputs, config }, props.children);
      };
      return {
        component: Component
      };
    }
  });
}
((createRouterExtension2) => {
  createRouterExtension2.componentDataRef = RouterBlueprint.dataRefs.component;
})(createRouterExtension || (createRouterExtension = {}));

export { createRouterExtension };
