// @backstage/frontend-plugin-api/dist/extensions/createSignInPageExtension.esm.js
import React, { lazy } from 'react';
import { ExtensionBoundary } from '../components/ExtensionBoundary.esm.js';
import '../wiring/coreExtensionData.esm.js';
import { createExtension } from '../wiring/createExtension.esm.js';
import '../blueprints/ApiBlueprint.esm.js';
import '../blueprints/AppRootElementBlueprint.esm.js';
import '../blueprints/AppRootWrapperBlueprint.esm.js';
import '../blueprints/IconBundleBlueprint.esm.js';
import '../blueprints/NavItemBlueprint.esm.js';
import '../blueprints/NavLogoBlueprint.esm.js';
import '../blueprints/PageBlueprint.esm.js';
import '../blueprints/RouterBlueprint.esm.js';
import { SignInPageBlueprint } from '../blueprints/SignInPageBlueprint.esm.js';
import '../blueprints/ThemeBlueprint.esm.js';
import '../blueprints/TranslationBlueprint.esm.js';

function createSignInPageExtension(options) {
  return createExtension({
    kind: "sign-in-page",
    namespace: options?.namespace,
    name: options?.name,
    attachTo: options.attachTo ?? { id: "app/root", input: "signInPage" },
    configSchema: options.configSchema,
    inputs: options.inputs,
    disabled: options.disabled,
    output: {
      component: createSignInPageExtension.componentDataRef
    },
    factory({ config, inputs, node }) {
      const ExtensionComponent = lazy(
        () => options.loader({ config, inputs }).then((component) => ({ default: component }))
      );
      return {
        component: (props) => /* @__PURE__ */ React.createElement(ExtensionBoundary, { node, routable: true }, /* @__PURE__ */ React.createElement(ExtensionComponent, { ...props }))
      };
    }
  });
}
((createSignInPageExtension2) => {
  createSignInPageExtension2.componentDataRef = SignInPageBlueprint.dataRefs.component;
})(createSignInPageExtension || (createSignInPageExtension = {}));

export { createSignInPageExtension };
