// @backstage/frontend-plugin-api/dist/extensions/createNavItemExtension.esm.js
import { createSchemaFromZod } from '../schema/createSchemaFromZod.esm.js';
import '../wiring/coreExtensionData.esm.js';
import { createExtension } from '../wiring/createExtension.esm.js';
import { NavItemBlueprint } from '../blueprints/NavItemBlueprint.esm.js';

function createNavItemExtension(options) {
  const { routeRef, title, icon, namespace, name } = options;
  return createExtension({
    namespace,
    name,
    kind: "nav-item",
    attachTo: { id: "app/nav", input: "items" },
    configSchema: createSchemaFromZod(
      (z) => z.object({
        title: z.string().default(title)
      })
    ),
    output: {
      navTarget: NavItemBlueprint.dataRefs.target
    },
    factory: ({ config }) => ({
      navTarget: {
        title: config.title,
        icon,
        routeRef
      }
    })
  });
}
((createNavItemExtension2) => {
  createNavItemExtension2.targetDataRef = NavItemBlueprint.dataRefs.target;
})(createNavItemExtension || (createNavItemExtension = {}));

export { createNavItemExtension };
